schemes/radiation_utils/solar_irradiance_data.F90

```diff
@@ -0,0 +1,404 @@
+
+       we(:nbins)  = lambda(:nbins) - 0.5_kind_phys*dellam(:nbins)
+       we(nbins+1) = lambda(nbins)  + 0.5_kind_phys*dellam(nbins)
+       do idx = 1,nbins
+          irrad_fac(idx) = 1.e-3_kind_phys                  ! mW/m2/nm --> W/m2/nm
+          etf_fac(idx)   = 1.e-16_kind_phys*lambda(idx)*fac ! mW/m2/nm --> photons/cm2/sec/nm
+       enddo
+       if(has_ref_spectrum) then
+          ssi_ref = ssi_ref * 1.e-3_kind_phys           ! mW/m2/nm --> W/m2/nm
+       endif
+    endif
+
+    deallocate(lambda)
+    deallocate(dellam)
+
+    ! need to force data loading when the model starts at a time =/ 00:00:00.000
+    ! -- may occur in restarts also
+    call solar_irradiance_data_run(irrad_file_path, nbins, nbinsp, has_spectrum, do_spectral_scaling, &
+            sol_irrad, we, sol_tsi, errmsg, errflg)
+    if (errflg /= 0) then
+       return
+    end if
+    initialized = .true.
+
+  end subroutine solar_irradiance_data_init
+
+!-------------------------------------------------------------------------------
+!-------------------------------------------------------------------------------
+
+!> \section arg_table_solar_irradiance_data_run Argument Table
+!! \htmlinclude solar_irradiance_data_run.html
+!!
+  subroutine solar_irradiance_data_run(irrad_file_path, nbins, nbinsp, has_spectrum, do_spectral_scaling, &
+                sol_irrad, we, sol_tsi, errmsg, errflg)
+     use ccpp_io_reader,   only: abstract_netcdf_reader_t, create_netcdf_reader_t
+     ! Arguments
+     character(len=*),   intent(in)    :: irrad_file_path
+     real(kind_phys),    intent(in)    :: we(:)                 ! wavelength endpoints
+     integer,            intent(in)    :: nbins                 ! number of bins
+     integer,            intent(in)    :: nbinsp                ! number of bins plus one
+     logical,            intent(in)    :: has_spectrum
+     logical,            intent(in)    :: do_spectral_scaling     ! flag to do spectral scaling
+     real(kind_phys),    intent(out)   :: sol_tsi               ! total solar irradiance
+     real(kind_phys),    intent(out)   :: sol_irrad(:)          ! solar irradiance
+     character(len=512), intent(out)   :: errmsg
+     integer,            intent(out)   :: errflg
+
+     ! Local variables
+     integer  :: idx, index, nt
+     integer  :: offset(2), count(2)
+     integer, allocatable :: itsi(:)
+     logical  :: read_data
+     real(kind_phys) :: data(nbins)
+     integer  :: ierr
+     real(kind_phys) :: delt
+     class(abstract_netcdf_reader_t), pointer :: file_reader
+
+     character(len=*), parameter :: subname = 'solar_irradiance_data_run: '
+
+     ! Initialize error variables
+     errflg = 0
+     errmsg = ''
+
+     if ( fixed_scon ) return
+     if ( time_coord%fixed .and. initialized ) return
+
+     index = -1
+
+     read_data = time_coord%read_more() .or. .not.initialized
+     call time_coord%advance()
+
+     if ( read_data ) then
+        file_reader => create_netcdf_reader_t()
```

Comment:
I am wondering if instead of constantly re-creating and destroying the `file_reader` object, we just make it a private module-level variable, and then only create it once during the `register` phase?  This might require some additional logic in the `init` phase to deallocate it once we have read the data for the "fixed" time configuration, but otherwise should make the code cleaner (and require less overall allocation/deallocation statements).  Of course happy to hear otherwise if you disagree!

Reply:
@nusbaume question for you! I agree with you that promoting the variable to the module level seems best, but:

Why does the reader need to be deallocated in the init phase?